File: data/models/Request.js
// models/Request.js
const mongoose = require( 'mongoose' );
const BaseSchema = require( './BaseSchema' );

const RequestSchema = new mongoose.Schema( {
  user:   { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  params: {
    songs_min_year_created: Number,
    songs_length_min_secs:  Number,
    songs_length_max_secs:  Number,
    language:               String,
    genre:                  String,
    lastfm_listeners_max:   Number,
    lastfm_listeners_min:   Number,
    lastfm_playcount_max:   Number,
    lastfm_playcount_min:   Number,
    spotify_followers_max:  Number,
    spotify_followers_min:  Number,
    niche_level:            String,
    lastfm_likeness_min:    Number,
    playlist_length:        Number,
  },
  playlist_generated: { type: mongoose.Schema.Types.ObjectId, ref: 'Playlist' },
} );

RequestSchema.add( BaseSchema );

module.exports = mongoose.model( 'Request', RequestSchema );

File: backend/src/models/Request.py
from mongoengine import EmbeddedDocument, EmbeddedDocumentField, StringField, IntField, ReferenceField
from models.BaseSchema import BaseSchema

class Params(EmbeddedDocument):
    songs_min_year_created = IntField()
    songs_length_min_secs  = IntField(required=True)
    songs_length_max_secs  = IntField(required=True)
    language               = StringField(required=True)
    genre                  = StringField(required=True)
    niche_level            = StringField(required=True)
    playlist_length        = StringField(required=True)

class Request(BaseSchema):
    user                = ReferenceField('users', required=True)
    params              = EmbeddedDocumentField(Params, required=True)
    playlist_generated  = ReferenceField('playlists')

    meta = {
        'collection': 'requests'
    }
